### ISSUE TITLE
[Vulkan/Beetle PSX HW] RE2 - Dual Shock Ver. - Texture warping error on Leon's face

### ISSUE THREAD

**Issue by cheds89** [libretro/beetle-psx-libretro#407]
Hi,

This is another visual bug I've noticed with Resident Evil 2 - Dual Shock Ver. with the Beetle PSX HW core using the Vulkan driver -

![resident evil 2 - dual shock ver usa -180813-195024](https://user-images.githubusercontent.com/42318042/44025556-cff7bf66-9f33-11e8-9949-8753d8fb6fb6.png)

The texture warping on Leon's face, particularly the eye area, seems really wrong to me. 

A comparison with the same scene using the Beetle PSX core (Software) shows how it should look -

![resident evil 2 - dual shock ver usa -180813-194023](https://user-images.githubusercontent.com/42318042/44025647-068b4e8a-9f34-11e8-90c3-2ff7732862b0.png)

Screenshots are taken at 4x GPU Internal Resolution.

I hope these are helpful.

Thanks!

**Comment by twinaphex** [comment]
Why is everything blue in your screenshot though? It makes it a bit hard to make any kind of comparisons.

Also, I think the texture warping on his face is just as a result of PGXP being disabled. The PlayStation suffers from lack of integer texture coordinates, so that and the lack of perspective correct texturing makes it that textures can warp. Add to that lack of subpixel precision and you have polygons too that wobble.

What you can try is setting PGXP Internal mode to ' Memory only'. With Resident Evil 1 this greatly helps with the facial warping/distortion.

So basically, I don't think this is an emulator bug, you are just running into one of the limitations of the PSX hardware. Luckily, we have additional features in the emulator that can mitigate these issues quite successfully nowadays, so try PGXP out.

**Comment by hizzlekizzle** [comment]
The red and blue channels are swapped on vulkan screenshots (on some GPUs but not others, maybe??)

Here's what it looks like with those channels swapped back in photoshop:
![untitled-1](https://user-images.githubusercontent.com/2805586/44033754-f842242e-9ed0-11e8-9b01-53787aac5dad.png)

**Comment by cheds89** [comment]
Thanks for both your replies.

Yep, the screenshots seem to have their blue and red channels reversed when using the Vulkan driver. Thanks for fixing hizzlekizzle. I also found it mentioned here - 

https://forums.libretro.com/t/red-blue-color-shift-in-screenshots-taken-with-vulkan-driver/10565

The issue is definitely fixed when using the PGXP correction features and it works really nicely. However, the texture warping displayed **without** the corrections is still definitely wrong compared to the Beetle Non-HW core.

I took 2 more screenshots in the same position. The first is in Beetle PSX (Non-HW) core with Vulkan driver -

![resident evil 2 - dual shock ver usa -180813-233130](https://user-images.githubusercontent.com/42318042/44034898-94c73fd4-9f51-11e8-92d3-e3a91fe81a41.png)

The second is in Beetle PSX HW core with Vulkan driver -

![resident evil 2 - dual shock ver usa -180813-232720](https://user-images.githubusercontent.com/42318042/44035009-d10d8cc8-9f51-11e8-9355-c7783896cccd.png)

I wish I could illustrate the difference with video but I'm having trouble capturing my screen, sorry.

Thanks.

**Comment by DukeJP** [comment]
With PGXP activated, vulkan renderer at 2x
![resident evil 2 - dual shock-180813-114150](https://user-images.githubusercontent.com/24689994/44039677-f28a3b56-9eef-11e8-9d01-4a9898858171.png)

But i find #406 another problem (second post)

**Comment by iCatButler** [comment]
There shouldn't be any difference between the distortion on models between software and hardware renderers when PGXP is disabled.

I think I see what you mean though. There's a sort of texture seam running across his face in the Vulkan pics. Does it snap into place only when you align the character on screen in a certain way or does it happen all the time?

Also, does it still happen if you set the video driver to GL?

@twinaphex it could be line rendering fix for Dark Forces and such, if that's still set to aggressive then it could be finding false candidates on the character's face and turning them into quads.

**Comment by iCatButler** [comment]
Those square edges are definitely created by the line rendering code I mentioned. It detects triangles with a specific size and orientation that would become solid pixel wide lines on a real PS1 and forces them into rendering as quads. It's needed to fix a handful of games that render parts of their scene using triangles to draw lines.

On 3D games without PGXP enabled there's no way for the code to tell if a triangle is one of these lines or part of a 3D model that just happens to meet the criteria after being transformed to screen space.

I believe twinaphex is working on selectively enabling this code, so it should only be a temporary glitch until that's implemented.

**Comment by cheds89** [comment]
Ok that's awesome. Thanks for the explanation! I hope this helped.

I guess I'll just keep this issue open until the fix is implemented?

**Comment by iCatButler** [comment]
Yeah, it's probably a good idea to leave it open in case anyone comes up with the same issue in the mean time.

**Comment by cheds89** [comment]
Cool will do.

Thanks

**Comment by iCatButler** [comment]
@cheds89 I believe this is now fixed in the current build, if you can give it a shot.

Just ensure the Line-to-Quad Hack option is set to default or off. The aggressive mode will still create these errors but it's only needed for a couple of specific games.

**Comment by cheds89** [comment]
Hey @iCatButler 

Thanks, I've had a quick look and the results look great. Here's some quick comparisons:

![re2](https://user-images.githubusercontent.com/42318042/44980771-2f757100-afb4-11e8-8118-ac345355138a.png)

![re2-2](https://user-images.githubusercontent.com/42318042/44980666-df96aa00-afb3-11e8-9840-41b1843a0578.png)

If it looks good to you, let me know and I'll close this issue.
